dataRetriever/resolvers/genericBlockBodyResolver.go

```diff
@@ -141,89 +140,88 @@ func (gbbRes *GenericBlockBodyResolver) RequestDataFromHashArray(hashes [][]byte
 }
 
 // GetMiniBlocks method returns a list of deserialized mini blocks from a given hash list either from data pool or from storage
-func (gbbRes *GenericBlockBodyResolver) GetMiniBlocks(hashes [][]byte) block.MiniBlockSlice {
-	miniBlocks := gbbRes.getMiniBlocks(hashes)
-	if miniBlocks == nil {
-		return nil
-	}
-
-	mbLength := len(hashes)
-	expandedMiniBlocks := make(block.MiniBlockSlice, mbLength)
-
-	for i := 0; i < mbLength; i++ {
-		mb := &block.MiniBlock{}
-		err := gbbRes.marshalizer.Unmarshal(mb, miniBlocks[i])
+func (gbbRes *GenericBlockBodyResolver) GetMiniBlocks(hashes [][]byte) (block.MiniBlockSlice, [][]byte) {
+	marshalizedMiniBlocks, missingMiniBlocksHashes := gbbRes.getMiniBlocks(hashes)
+	miniBlocks := make(block.MiniBlockSlice, 0)
 
+	for hash, marshalizedMiniBlock := range marshalizedMiniBlocks {
+		miniBlock := &block.MiniBlock{}
+		err := gbbRes.marshalizer.Unmarshal(miniBlock, marshalizedMiniBlock)
 		if err != nil {
 			log.Debug(err.Error())
-			gbbRes.miniBlockPool.Remove(hashes[i])
-			err = gbbRes.miniBlockStorage.Remove(hashes[i])
+			gbbRes.miniBlockPool.Remove([]byte(hash))
+			err = gbbRes.miniBlockStorage.Remove([]byte(hash))
 			if err != nil {
 				log.Debug(err.Error())
 			}
 
-			return nil
+			missingMiniBlocksHashes = append(missingMiniBlocksHashes, []byte(hash))
+			continue
 		}
 
-		expandedMiniBlocks[i] = mb
+		miniBlocks = append(miniBlocks, miniBlock)
 	}
 
-	return expandedMiniBlocks
+	return miniBlocks, missingMiniBlocksHashes
 }
 
 // getMiniBlocks method returns a list of serialized mini blocks from a given hash list either from data pool or from storage
-func (gbbRes *GenericBlockBodyResolver) getMiniBlocks(hashes [][]byte) [][]byte {
-	miniBlocks := gbbRes.getMiniBlocksFromCache(hashes)
+func (gbbRes *GenericBlockBodyResolver) getMiniBlocks(hashes [][]byte) (map[string][]byte, [][]byte) {
+	marshalizedMiniBlocks, missingMiniBlocksHashes := gbbRes.getMiniBlocksFromCache(hashes)
+	if len(missingMiniBlocksHashes) == 0 {
+		return marshalizedMiniBlocks, missingMiniBlocksHashes
+	}
 
-	if miniBlocks != nil {
-		return miniBlocks
+	marshalizedMiniBlocksFromStorer, missingMiniBlocksHashes := gbbRes.getMiniBlocksFromStorer(missingMiniBlocksHashes)
```

Comment:
why would some of the miniblocks be already in store ? At rollback you get everything from store and put in miniblock pool. Is there an other edgecase where an unprocessed miniblock might get saved into the store and not restored ?

Reply:
This method is also called  from resolveBlockBodyRequest when someone else requested for a mini block. In this case, if the requested mini block is too old (when somebody is syncing) , the only place where it can still be found is in storer.